[macOS] Add support for extending background colors beyond fixed-position containers when obscured insets are set

whsieh opened the pull request:
#### 9f579371678508d1445d6b906bce3507522310f3
<pre>
[macOS] Add support for extending background colors beyond fixed-position containers when obscured insets are set
<a href="https://bugs.webkit.org/show_bug.cgi?id=289877">https://bugs.webkit.org/show_bug.cgi?id=289877</a>
<a href="https://rdar.apple.com/147177776">rdar://147177776</a>

Reviewed by Abrar Rahman Protyasha.

Add support for extending background colors beyond fixed-position containers, on sides with nonzero
obscured content insets.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _recalculateViewportSizesWithMinimumViewportInset:maximumViewportInset:throwOnInvalidInput:]):
(-[WKWebView _coreFixedContainerEdges]):

Expose an internal getter for `WebCore::FixedContainerEdges`.

(-[WKWebView _updateFixedContainerEdges:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateFixedContentExtensionViews):

Add a new helper method to update extension views for fixed-position containers.

Canonical link: <a href="https://commits.webkit.org/292256@main">https://commits.webkit.org/292256@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe70a107af8078d1e620626c5a9ebcf5d7922ac5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95411 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4877 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100459 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45917 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15301 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72770 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30035 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98413 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11445 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86126 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53101 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11158 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45252 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81336 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3953 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102496 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22461 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16395 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81787 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22709 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82137 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81158 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20303 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25734 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3156 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15758 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22430 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27569 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22089 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25564 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23829 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6915,6 +6915,55 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
+            [extensionView setBackgroundColor:predominantColor.get() ?: [webView underPageBackgroundColor]];
+            [extensionView setFrame:computeFrame()];
+        }
+        [extensionView setHidden:!needsView];
```
I think there's an implicit `bool` to `BOOL` here?

whsieh replied:
So I _believe_ implicit conversion from `bool` to `BOOL` is generally fine — it only breaks in Catalyst if it's being done for the return value of a lambda or ObjC block, with no explicit return type.

aprotyas commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6915,6 +6915,55 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
+            if (!extensionView) {
+                extensionView = adoptNS([NSView new]);
+                [extensionView setWantsLayer:YES];
+                [extensionView layer].name = @"Fixed color extension fill";
```
I wonder if it will help if you also included the box side.

whsieh replied:
Oh, good idea! Will add the side.

whsieh committed:
[macOS] Add support for extending background colors beyond fixed-position containers when obscured insets are set
https://bugs.webkit.org/show_bug.cgi?id=289877
rdar://147177776

Reviewed by Abrar Rahman Protyasha.

Add support for extending background colors beyond fixed-position containers, on sides with nonzero
obscured content insets.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _recalculateViewportSizesWithMinimumViewportInset:maximumViewportInset:throwOnInvalidInput:]):
(-[WKWebView _coreFixedContainerEdges]):

Expose an internal getter for `WebCore::FixedContainerEdges`.

(-[WKWebView _updateFixedContainerEdges:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateFixedContentExtensionViews):

Add a new helper method to update extension views for fixed-position containers.

Canonical link: https://commits.webkit.org/292256@main
